### PR TITLE
Implement triangular

### PR DESCRIPTION
**Pull request by BruceDai**
@fdwr @huningxin PTAL, thanks.

### CODE REVIEW

File: src/lib/validate-input.js [hunk 1]
```diff
@@ -444,3 +444,17 @@ export function validateGatherParams(input, indices, {axis = 0} = {}) {
...
+export function validateTriangularParams(input, {diagonal = 0} = {}) {
+  const inputRank = input.rank;
+  if (inputRank !== 2) {
+    throw new Error('The input should be a 2-D tensor.');
+  }
+  const [i, j] = input.shape;
+  if (i !== j) {
+    throw new Error('The input should be a 2-D tensor of [N, N] shape.');
+  }
+  if (diagonal >= i || diagonal <= -i) {
```
Comment (BruceDai) [inline src/lib/validate-input.js:457]: Spec doesn't give the restrict for this **diagonal** . Please correct me, if I understand wrongly. Thanks.

File: src/lib/validate-input.js [hunk 2]
```diff
@@ -444,3 +444,17 @@ export function validateGatherParams(input, indices, {axis = 0} = {}) {
...
+export function validateTriangularParams(input, {diagonal = 0} = {}) {
+  const inputRank = input.rank;
+  if (inputRank !== 2) {
+    throw new Error('The input should be a 2-D tensor.');
+  }
+  const [i, j] = input.shape;
+  if (i !== j) {
+    throw new Error('The input should be a 2-D tensor of [N, N] shape.');
+  }
+  if (!Number.isInteger(diagonal) || diagonal >= i || diagonal <= -i) {
```
Comment (BruceDai) [inline src/lib/validate-input.js:457]: Spec doesn't give the restrict for this diagonal . Please correct me, if I understand wrongly. Thanks.

Comment (fdwr) [reply src/lib/validate-input.js:457]: > Spec doesn't give the restrict for this diagonal . 

Correct - no restriction here. If diagonal is beyond the dimensions of the output tensor, then the output will simply be fully zero or fully copied.

Comment (BruceDai) [reply src/lib/validate-input.js]: Updated by removing this restriction. Please take another look, thanks.

File: test/triangular_test.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+          upper: false,
+          diagonal: -1,
+        },
+    );
+  });
+
+  it('triangular fully zero upper=false diagonal=-4', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
```
Comment (fdwr) [inline test/triangular_test.js:234]: All the shapes are square. We should include at least one asymmetric case (different height and width) to ensure we don't have any bugs where we accidentally swap rows for columns (e.g. `shape: [4, 3]`.)

File: test/triangular_test.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
```
Comment (fdwr) [inline test/triangular_test.js:242]: We should also include one with a batch count other than 0. 🤨 It looks like a spec typo that I missed where it says `triangular` takes exactly 2D tensors, but instead it should say the input is *at least* 2D. You can see other ML libraries just treat each 2D subsection as a separate batch:

https://numpy.org/doc/stable/reference/generated/numpy.triu.html
https://pytorch.org/docs/stable/generated/torch.tril.html
https://onnx.ai/onnx/operators/onnx__Trilu.html

So shape [4,3,5,7] would apply the triangular masking to the last two dimensions. Let's create a WebNN issue...

**Review by fdwr — COMMENTED** [review]
Let's add 2 more cases, but everything you already have LGTM.

File: test/triangular_test.js (hunk 3 above)
Comment (BruceDai) [reply test/triangular_test.js:234]: Thanks @fdwr .
Last I referred to https://en.wikipedia.org/wiki/Triangular_matrix:
```
In mathematics, a triangular matrix is a special kind of [square matrix](https://en.wikipedia.org/wiki/Square_matrix).
``
OK, I will update it without restricting square matrix.

File: test/triangular_test.js (hunk 4 above)
Comment (BruceDai) [reply test/triangular_test.js:242]: > Let's create a WebNN issue...

Done, please see https://github.com/webmachinelearning/webnn/issues/494, thanks.

File: test/triangular_test.js (hunk 3 above)
Comment (fdwr) [reply test/triangular_test.js:234]: I've read other definitions like "An m×n matrix A is called upper triangular if all entries lying below the diagonal entries are zero" which imply m and n can be distinct, but it's certainly that that square matrices (a subset) satisfy that too. ONNX doesn't care. Even Wolfram Alpha accepts the following fine `MatrixForm[UpperTriangularize[{{1, 2, 3, 4}, {4, 5, 6, 7}, {7, 8, 9, 10}}]]`.

> OK, I will update it without restricting square matrix.

👍

**Comment by BruceDai** [comment]
@fdwr I've updated commit to support at least 2D tensor, please take another look, thanks.

**Review by fdwr — APPROVED** [review]
👍

**Review by huningxin — APPROVED** [review]

**Pull request merged**